feat: Adds PKCE authentication for GitLab closes #5236

something-something-something opened the pull request:
<!--
Thanks for submitting a pull request!

Please make sure you've read and understood our contributing guidelines.

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx", where #xxxx is the issue number.

Please provide enough information so that others can review your pull request.
The first two fields are mandatory:
-->

**Summary**

<!--
Explain the **motivation** for making this change.
What existing problem does the pull request solve?
-->
closes #5236
This adds PKCE authentication for GitLab as the existing authentication method (implicit)  is going to be deprecated. 
see https://gitlab.com/gitlab-org/gitlab/-/issues/288516 and https://docs.gitlab.com/ce/api/oauth2.html#authorization-code-with-proof-key-for-code-exchange-pkce

Configuration for use is as follows:

```
backend:
  name: gitlab
  repo: user/repo
  auth_type: pkce
  app_id: someapid898909439034904390490
```

I copied the existing implicit oauth implementation and changed it to perform pkce authentication instead. 

As such some functions and other code like createNonce()  and  validateNonce() are the same as the ones in the implicit implementation and should be moved to a separate file. Where should they go?

**Test plan**

<!--
Demonstrate the code is solid.
Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI.
-->
None yet. Where do tests go for authentication?

**A picture of a cute animal (not mandatory but encouraged)**

erezrokah reviewed (changes requested):
Thanks , can we add some information about setting this up in the docs https://www.netlifycms.org/docs/gitlab-backend/#client-side-implicit-grant-gitlab
and maybe add a note about the deprecation of implicit auth?

You can create a new file called `utils.js` file under `netlify-cms-lib-auth/src` and move shared code to it.

For tests, we [pre-recorded browser traffic](https://github.com/netlify/netlify-cms/tree/master/cypress/fixtures) to test integrations with different providers.

I don't think we test implicit auth though.

something-something-something commented:
Okay I moved the shared functions to utils.js and added docs. 

For docs should implicit just say something like:

Follow the instructions for PKCE but replace 
```yaml
  auth_type: pkce
```
with
```yaml
  auth_type: implicit
```

> I don't think we test implicit auth though.

Is it okay if I don't write test for pkce?

erezrokah reviewed (approved):
This is great @something-something-something.

I made some changes in https://github.com/netlify/netlify-cms/pull/5239/commits/9c7dbb600986cc87a5995e2fad0ec7c4528a3bd4 but they are mostly cosmetic.

If you're happy with https://github.com/netlify/netlify-cms/pull/5239/commits/9c7dbb600986cc87a5995e2fad0ec7c4528a3bd4 I think this is ready.

Since adding tests for this is quite tricky, I don't think this should block the PR.

erezrokah commented on packages/netlify-cms-backend-gitlab/src/AuthenticationPage.js:
```diff
@@ -34,11 +42,12 @@ export default class GitLabAuthenticationPage extends React.Component {
 
-    if (authType === 'implicit') {
-      this.auth = new ImplicitAuthenticator({
+    if (clientSideAuthenticators[authType]) {
```
Avoids some code duplication

erezrokah commented on packages/netlify-cms-lib-auth/src/pkce-oauth.js:
```diff
@@ -3,33 +3,49 @@ import trimEnd from 'lodash/trimEnd';
+  const encoder = new TextEncoder();
+  const data = encoder.encode(text);
+  const digest = await window.crypto.subtle.digest('SHA-256', data);
+  const sha = String.fromCharCode(...new Uint8Array(digest));
```
`Uint8Array` is already an array and `String.fromCharCode` accepts a sequence

erezrokah commented on packages/netlify-cms-lib-auth/src/pkce-oauth.js:
```diff
@@ -3,33 +3,49 @@ import trimEnd from 'lodash/trimEnd';
+  // excludes _~ as if included would cause an uneven distribution as char.length would no longer be a factor of 256
+  const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-.';
+  const randomValues = window.crypto.getRandomValues(new Uint8Array(128));
+  return randomValues
```
Same, `Uint8Array` is already an array

something-something-something replied:
I think You need Array.from here

See the difference in the following (can be pasted into a browser):


The following only produces numbers:

```js
(()=>{
	const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-.';
	const randomValues = window.crypto.getRandomValues(new Uint8Array(128));
	return randomValues.map((val) => {
		return chars[val % chars.length];
	}).join('');
})();
```

Will produce proper output:
```js
(()=>{
	const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-.';
	const randomValues = Array.from(window.crypto.getRandomValues(new Uint8Array(128)));
	return randomValues.map((val) => {
		return chars[val % chars.length];
	}).join('');
})();

```

something-something-something replied:
The issue is the Uint8Array.map returns an Uint8Array. not a regular array so all the strings are converted to numbers. Strings that can't convert to numbers become 0.

erezrokah replied:
Nice one thanks for explaining!

erezrokah commented on packages/netlify-cms-lib-auth/src/pkce-oauth.js:
```diff
@@ -58,24 +74,17 @@ export default class PkceAuthenticator {
-      //
-      document.location.assign(authURL.href);
-    });
+    const codeVerifier = createCodeVerifier();
```
Make it easier to understand the flow

erezrokah commented on packages/netlify-cms-lib-auth/src/pkce-oauth.js:
```diff
@@ -3,33 +3,49 @@ import trimEnd from 'lodash/trimEnd';
     .join('');
 }
 
-function createOrGetCodeVerifier() {
```
I separated this function to make it more explicit.
Before it seems that creating a code verifier in `completeAuth` could be consider as a valid use case (am I assuming correctly that it is not?)

something-something-something commented:
Thanks! I added back Array.from to generateVerifierCode() as it does in fact need to be there. All the other changes are good with me. I am happy with https://github.com/something-something-something/netlify-cms/commit/aaedd995e669115257f4611ae3aee8d5148cecc2 being merged.
